README.md
## React Native FadingSlides Component

[![DeepScan Grade](https://deepscan.io/api/projects/1122/branches/2515/badge/grade.svg)](https://deepscan.io/dashboard/#view=project&pid=1122&bid=2515)

Simple looped fading slides carousel for React Native.

![alt tag](http://i.giphy.com/l41lR24WgEwYTaHzW.gif)

### Installation

```bash
npm install --save react-native-fading-slides
```

### Properties

```
fadeDuration={500} // Milliseconds for slide fade
stillDuration={1000} // Milliseconds for slide still
height={1000} // Set the slides component height
slides={slidesList} // Set the slides list
startAnimation={true} // Start or stops animation
```

### Slides Properties

```
title={"Title"} // Slide's title
titleColor={"#fff"} // Slide's title color
subtitle={"Subtitle"} // Slide's subtitle
subtitleColor={"#fff"} // Slide's subtitle color
image={require('image!filename')} // Slide's image
imageWidth={1000} // Slide's image width
imageHeight={1000} // Slide's image height
```

### Usage Example

```javascript
import FadingSlides from 'react-native-fading-slides';

const slides = [
  {
    image: require('image!squared-image'),
    imageWidth: 100,
    imageHeight: 100,
    title: 'Hello World',
    subtitle: 'This is a beautiful world',
    titleColor: '#fff',
    subtitleColor: '#fff',
  },
  {
    image: require('image!wide-image'),
    imageWidth: 200,
    imageHeight: 100,
    title: 'Bye World',
    subtitle: 'This is a see you soon',
    titleColor: '#fff',
    subtitleColor: '#fff',
  }
];

//...

render() {
  return (
    <View>
      <FadingSlides
        slides={slides}
        fadeDuration={1200}
        stillDuration={2000}
        height={500}
        startAnimation={true}
      />
    </View>
  );
};
```


CHANGELOG.md
### 0.1.0

* Initial component

### 0.1.1

* Add keywords for react.parts component submission

### 0.1.2

* Fix package dependencies with react-timer-mixin and react-tween-state

### 0.1.4

* Remove peerDependencies and update react import (now compatible with recent
  react-native versions)

### 0.2.0

* Use Animated API instead of TimerMixin and TweenState

### 0.3.0

* Fixes memory leak and adds startAnimation property

### 0.3.1

* Fixes issue with slide change


index.js
import React from 'react';
import { StyleSheet, Text, Image, View, Animated, Easing } from 'react-native';
import PropTypes from 'prop-types';

const MINIMUM_DELAY = 100;

export default class FadingSlides extends React.Component {
  state = {
    opacity: new Animated.Value(1),
    currentIndex: 0
  };

  componentDidMount() {
    if (this.props.startAnimation) {
      this._wait(this._hide);
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.startAnimation) {
      this._wait(this._hide);
    }
  }

  _animate = (targetValue, cb) => {
    Animated.timing(this.state.opacity, {
      toValue: targetValue,
      easing: Easing.circle,
      duration: this.props.fadeDuration,
      useNativeDriver: true
    }).start(() => this._wait(cb));
  };

  _wait = cb => {
    Animated.delay(this.props.stillDuration).start(cb);
  };

  _show = () => {
    this._animate(1, this._hide);
  };

  _hide = () => {
    this._animate(0, this._changeSlide);
  };

  _changeSlide = () => {
    let index = this.state.currentIndex + 1;
    index = index < this.props.slides.length ? index : 0;
    if (this.props.startAnimation) {
      this.setState({ currentIndex: index }, this._show);
      counter = index;
    }
  };

  render() {
    const slide = this.props.slides[this.state.currentIndex];
    return (
      <Animated.View
        style={[
          styles.slide,
          { height: this.props.height, opacity: this.state.opacity }
        ]}
      >
        <View style={styles.info} />
        <Image
          style={{ width: slide.imageWidth, height: slide.imageHeight }}
          source={slide.image}
        />
        <View style={styles.info}>
          <Text style={[styles.title, { color: slide.titleColor }]}>
            {slide.title}
          </Text>
          <Text style={[styles.subtitle, { color: slide.subtitleColor }]}>
            {slide.subtitle}
          </Text>
        </View>
      </Animated.View>
    );
  }
}

const styles = StyleSheet.create({
  slide: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    flexDirection: 'column'
  },
  info: {
    flex: 1,
    justifyContent: 'flex-end',
    alignItems: 'center',
    flexDirection: 'column'
  },
  title: {
    fontSize: 22,
    fontWeight: '700'
  },
  subtitle: {
    fontSize: 20,
    paddingBottom: 30
  }
});

FadingSlides.propTypes = {
  startAnimation: PropTypes.bool,
  stillDuration: PropTypes.number,
  fadeDuration: PropTypes.number,
  height: PropTypes.number,
  slides: PropTypes.arrayOf(
    PropTypes.shape({
      image: PropTypes.number,
      imageWidth: PropTypes.number,
      imageHeight: PropTypes.number,
      title: PropTypes.string,
      subtitle: PropTypes.string,
      titleColor: PropTypes.string,
      subtitleColor: PropTypes.string
    })
  )
};

FadingSlides.defaultProps = {
  startAnimation: false,
  stillDuration: MINIMUM_DELAY,
  fadeDuration: MINIMUM_DELAY
};


example/App.js
import React from 'react';
import { StyleSheet, View, Dimensions } from 'react-native';

import FadingSlides from 'react-native-fading-slides';

const { width, height } = Dimensions.get('window');

const slides = [
  {
    image: require('./img/javascript.png'),
    imageWidth: width - width * 0.3,
    imageHeight: width - width * 0.3,
    title: 'JavaScript',
    subtitle: 'The definitive language',
    titleColor: '#fff',
    subtitleColor: 'yellow',
  },
  {
    image: require('./img/react.png'),
    imageWidth: width - width * 0.3,
    imageHeight: width - width * 0.3,
    title: 'ReactJS',
    subtitle: 'JavaScript coolest library',
    titleColor: '#fff',
    subtitleColor: 'cyan',
  },
];

export default class App extends React.Component {
  render() {
    return (
      <View style={styles.container}>
        <FadingSlides slides={slides} fadeDuration={1200} stillDuration={2000} height={height} />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#000',
  },
});


fading-slides.component.js
import React from "react";
import { Animated, StyleSheet, View, Text } from "react-native";

import TimerMixin from "react-timer-mixin";
import tweenState from "react-tween-state";

let MINIMUM_DELAY = 100;

export default class FadingSlides extends React.Component {
  mixins: [TimerMixin, tweenState.Mixin];

  state = {
    opacity: 0,
    currentIndex: 0
  };

  componentDidMount() {
    this._fadeOpacity();
  }

  _fadeOpacity() {
    this.tweenState("opacity", {
      easing: tweenState.easingTypes.easeInCirc,
      duration: this.props.fadeDuration,
      beginValue: this.state.opacity === 0 ? 0 : 1,
      onEnd: this._getNextSlide,
      endValue: this.state.opacity === 0 ? 1 : 0
    });
  }

  _getNextSlide() {
    if (this.state.opacity === 0) {
      var index = this.state.currentIndex + 1;
      index = index < this.props.slides.length ? index : 0;
      this.setState({ currentIndex: index });
      this.setTimeout(() => {
        this._fadeOpacity();
      }, MINIMUM_DELAY);
    } else {
      this.setTimeout(() => {
        this._fadeOpacity();
      }, this.props.stillDuration || MINIMUM_DELAY);
    }
  }

  render() {
    var slide = this.props.slides[this.state.currentIndex];
    return (
      <View
        style={[
          styles.slide,
          {
            height: this.props.height,
            opacity: this.getTweeningValue("opacity")
          }
        ]}
      >
        <View style={styles.info} />
        <Animated.Image
          style={{ width: slide.imageWidth, height: slide.imageHeight }}
          source={slide.image}
        />
        <View style={styles.info}>
          <Text style={[styles.title, { color: slide.titleColor }]}>
            {slide.title}
          </Text>
          <Text style={[styles.subtitle, { color: slide.subtitleColor }]}>
            {slide.subtitle}
          </Text>
        </View>
      </View>
    );
  }
}

var styles = StyleSheet.create({
  slide: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
    flexDirection: "column"
  },
  info: {
    flex: 1,
    justifyContent: "flex-end",
    alignItems: "center",
    flexDirection: "column"
  },
  title: {
    fontSize: 22,
    fontWeight: "700"
  },
  subtitle: {
    fontSize: 20,
    paddingBottom: 30
  }
});


example/index.android.js
import React, { Component } from 'react';

import {
  AppRegistry,
  Dimensions,
  StyleSheet,
  Text,
  View,
} from 'react-native';

var { width, height } = Dimensions.get('window');

const FadingSlides = require('react-native-fading-slides');

const slides = [
  {
    image: require('./img/javascript.png'),
    imageWidth: width - (width * 0.3),
    imageHeight: width - (width * 0.3),
    title: 'JavaScript',
    subtitle: 'The definitive language',
    titleColor: '#fff',
    subtitleColor: 'yellow'
  },
  {
    image: require('./img/react.png'),
    imageWidth: width - (width * 0.3),
    imageHeight: width - (width * 0.3),
    title: 'ReactJS',
    subtitle: 'JavaScript coolest library',
    titleColor: '#fff',
    subtitleColor: 'cyan'
  }
];

class example extends Component {
  render() {
    return (
      <View style={styles.container}>
        <FadingSlides
          slides={slides}
          fadeDuration={1200}
          stillDuration={2000}
          height={height}
        />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#000',
  },
});

AppRegistry.registerComponent('example', () => example);
